vertical-pod-autoscaler/pkg/recommender/routines/safety_margin_modifier_post_processor.go

```diff
@@ -0,0 +1,221 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package routines
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+	"k8s.io/klog/v2"
+)
+
+type SafetyMarginModifierPostProcessor struct {
+	DefaultSafetyMarginFactor float64
+}
+
+type modifierFunction int64
+
+const (
+	undefined modifierFunction = iota
+	linear
+	log
+	affine
+	exponential
+)
+
+type safetyMarginFunction struct {
+	Function   modifierFunction
+	Parameters []float64
+}
+
+type safetyMarginModifier map[v1.ResourceName]safetyMarginFunction
+
+const (
+	vpaPostProcessorSafetyMarginModifierSuffix = "_safetyMarginModifier"
+)
+
+var _ RecommendationPostProcessor = &SafetyMarginModifierPostProcessor{}
+
+// Process applies the Resource Ratio post-processing to the recommendation.
+func (r *SafetyMarginModifierPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, policy *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+	updatedRecommendation := recommendation.DeepCopy()
+	modifiers := readSafetyMarginModifierFromVPAAnnotations(vpa)
+
+	for _, containerRec := range updatedRecommendation.ContainerRecommendations {
+		modifier, found := modifiers[containerRec.ContainerName]
+
+		if !found {
+			continue
+		}
+		klog.Infof("Apply safety marging modifier on container %s in vpa %s/%s", containerRec.ContainerName, vpa.ID.Namespace, vpa.ID.VpaName)
+
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.LowerBound, modifier)
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.Target, modifier)
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.UpperBound, modifier)
+		r.applyModifier(vpa, containerRec.ContainerName, containerRec.UncappedTarget, modifier)
+	}
+
+	return updatedRecommendation
+}
+
+func (r *SafetyMarginModifierPostProcessor) applyModifier(vpa *model.Vpa, name string, containerRec v1.ResourceList, modifier safetyMarginModifier) {
+	for resourceName, rec := range containerRec {
+		resourceModifier, found := modifier[resourceName]
+		if !found {
+			continue
+		}
+
+		var newRec float64
+
+		// CPU requests are expressed in milli cores, which impacts the modification brough by
+		// the Log and Exponential modifiers. If the resource is CPU, scale it by 1000.
+		recCopy := rec.DeepCopy()
+		if resourceName == v1.ResourceCPU {
+			recCopy.SetScaled(rec.MilliValue(), 0)
```

Comment:
The issue is that the unit of CPU on K8s is the millicore, but the quantity is in cores, so the quantity has a scale of `-3`. This means that we have a high chance to have values below 1.
For linear and affine modifier, this has no impact on the recommendation, but for log and exponential, the result is not the same. In particular for `log`, this can cause the recommendation to be lower than the p95 usage, and can even lead to recommendation of 0 given that log diverges at 0. 

So imo, it makes more sense to scale it by 1000, to operated on integer millicore values, rather than float core values.

Reply:
Oh I see later you are using NewMilliQuantity with this value that is store in quantity with scale 0.... This is so confusing, it would be so simple to have a float that represent millicore everywhere (I continue to hate this Quantity object).

Code LGTM. The doc will have to mention that function use millicores and not cores